src/Symfony/Component/HttpKernel/MicroKernel.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel;
+
+use Symfony\Bundle\FrameworkBundle\Routing\RouteCollectionBuilder;
+use Symfony\Component\Config\Loader\LoaderInterface;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * A Kernel that provides configuration hooks.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+abstract class MicroKernel extends Kernel
+{
+    /**
+     * Add or import routes into your application.
+     *
+     *     $routes->import('config/routing.yml');
+     *     $routes->add('/admin', 'AppBundle:Admin:dashboard', 'admin_dashboard');
+     *
+     * @param RouteCollectionBuilder $routes
+     */
+    protected function configureRoutes(RouteCollectionBuilder $routes)
+    {
+        return $routes;
+    }
+
+    /**
+     * Configures container extensions.
+     *
+     * $c->loadFromExtension('framework', array(
+     *     'secret' => '%secret%'
+     * ));
+     *
+     * @param ContainerBuilder $c
+     * @param LoaderInterface  $loader
+     */
+    protected function configureExtensions(ContainerBuilder $c, LoaderInterface $loader)
+    {
+    }
+
+    /**
+     * Adds service definitions to the container.
+     *
+     * @param ContainerBuilder $c
+     * @param LoaderInterface  $loader
+     */
+    protected function configureServices(ContainerBuilder $c, LoaderInterface $loader)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function registerContainerConfiguration(LoaderInterface $loader)
+    {
+        $loader->load(function (ContainerBuilder $container) use ($loader) {
+            $container->loadFromExtension('framework', array(
+                'router' => array(
+                    'resource' => null,
+                    'type' => 'file',
+                ),
+            ));
+
+            $this->configureExtensions($container, $loader);
+            $this->configureServices($container, $loader);
+
+            $container->addObjectResource($this);
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function boot()
+    {
+        if (true === $this->booted) {
+            return;
+        }
+
+        parent::boot();
+
+        // 3.0 only
+        $this->getContainer()->get('router')->setResource(function () {
+            $loader = $this->getContainerLoader($this->getContainer());
+
+            return $this->configureRoutes(new RouteCollectionBuilder($loader))->build();
+        });
+    }
```

Comment:
This starts to feel a bit worse than the ServiceRouteLoader approach, which also had some supporters for other use-cases. If there were a way to pull this off in a nice way without any new code, that would be great. But otherwise, I like #15742